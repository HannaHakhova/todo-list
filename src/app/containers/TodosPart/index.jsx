import React, {Component} from 'react';
import * as R from 'ramda';
import {Segment, Header} from 'semantic-ui-react';
import Footer from 'components/Footer';
import TodoInput from 'components/TodoInput';
import TodoList from 'components/TodoList';
import {getFormattedDate} from 'utils';
import {bool, arrayOf, string, number, shape, func} from 'prop-types';

class TodosPart extends Component {
    constructor(props) {
        super(props);
        this.state = {
            activeItem: 'all',
            directionUp: true,
            todos: []
        };
    }

    componentWillReceiveProps(nextProps) {
        this.prepareTodos(nextProps.todos, nextProps.tags, nextProps.todosTags);
    }

    prepareTodos = (todos, tags, todosTags) => {
        const newTodos = todos.map(todo => {
            const todosTagsForCurrentTodo = todosTags.filter(f => f.taskId === todo.id);
            const tagsForCurrentTodo = todosTagsForCurrentTodo.map(currentTodosTags => ({...tags.filter(tag => tag.id === currentTodosTags.tagId)[0], bindingId: currentTodosTags.id}));
            return {...todo, tags: tagsForCurrentTodo};
        });
        this.setState({todos: newTodos.map(x => ({...x, createdDate: getFormattedDate(x.createdDate)}))});
    };

    handleActiveItemChange = activeItem => this.setState({activeItem});

    sortByDateAndCompleted = () => {
        const sortDirection = this.state.directionUp === true ? R.ascend : R.descend;
        const todos = R.sortWith([
            sortDirection(R.prop('createdDate')),
            sortDirection(R.prop('completed'))
        ])(this.state.todos);
        this.setState({todos, directionUp: !this.state.directionUp});
    }

    render() {
        return (
            <div>
                <br />
                <Segment size='mini' textAlign='left' padded='very'>
                    <Header textAlign='center' as='h2'>Todo list </Header>
                    <TodoInput handleSubmit={this.props.createTodo} />
                    <TodoList handleComplete={this.props.toggleCompleted} handleDelete={this.props.deleteTodo} todos={this.state.todos} handleDeleteTodoTag={this.props.deleteTodoTag} activeItem={this.state.activeItem} sortByDateAndCompleted={this.sortByDateAndCompleted} addTodoTag={this.props.addTodoTag} />
                    <Footer
                        todos={this.state.todos}
                        handleClearCompleted={this.props.clearCompleted}
                        handleActiveItemChange={this.handleActiveItemChange} />
                </Segment>
            </div>
        );
    }
}

TodosPart.propTypes = {
    todos: arrayOf(shape({
        id: string.isRequired,
        title: string.isRequired,
        completed: bool.isRequired,
        createdDate: string.isRequired}).isRequired).isRequired,
    todosTags: arrayOf(shape({
        id: number.isRequired,
        taskId: string.isRequired,
        tagId: string.isRequired}).isRequired).isRequired,
    tags: arrayOf(shape({
        id: string.isRequired,
        title: string.isRequired}).isRequired).isRequired,
    createTodo: func.isRequired,
    deleteTodo: func.isRequired,
    toggleCompleted: func.isRequired,
    clearCompleted: func.isRequired,
    deleteTodoTag: func.isRequired,
    addTodoTag: func.isRequired
};

export default TodosPart;
