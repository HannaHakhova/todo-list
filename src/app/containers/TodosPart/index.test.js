/* eslint-disable react/jsx-filename-extension */

import React from 'react';
import {shallow} from 'enzyme';
import TodosPart from './index';

jest.mock('api', () => ({
    getAllData: () => new Promise(resolve => {
        const response = {
            data: [
                {
                    id: '94f3ab7b-f256-4d40-8f98-d6a7c625d816',
                    title: 'Berkshire',
                    completed: true,
                    createdDate: '6/12/2017'
                },
                {
                    id: 'b6fadd6b-2977-4797-8773-ba61da28311b',
                    title: 'Administrator',
                    completed: true,
                    createdDate: '5/12/2017'
                }
            ]
        };
        resolve(response);
    }),
    createData: () => {},
    deleteData: () => {},
    updateData: () => {}
}));

jest.mock('utils', () => ({
    getFormattedDate: () => new Date('Thu Apr 19 2018 12:18:16 GMT+0300 (FLE Daylight Time)'),
    getDate: () => {},
    getApiDate: () => {}
}));

describe('TodosPart', () => {
    it('matches expected snapshot', () => {
        const wrapper = shallow(<TodosPart />);
        expect(wrapper).toMatchSnapshot();
    });

    it('matches expected state ', done => {
        const wrapper = shallow(<TodosPart />);
        const todos = [
            {
                id: '94f3ab7b-f256-4d40-8f98-d6a7c625d816',
                title: 'Berkshire',
                completed: true,
                createdDate: new Date('Thu Apr 19 2018 12:18:16 GMT+0300 (FLE Daylight Time)')
            },
            {
                id: 'b6fadd6b-2977-4797-8773-ba61da28311b',
                title: 'Administrator',
                completed: true,
                createdDate: new Date('Thu Apr 19 2018 12:18:16 GMT+0300 (FLE Daylight Time)')
            }];

        setTimeout(() => {
            expect(wrapper.state().todos).toEqual(todos);
            done();
        }, 100);
    });
});