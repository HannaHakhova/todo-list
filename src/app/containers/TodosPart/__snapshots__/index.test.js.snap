// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodosPart matches expected snapshot 1`] = `
<div>
  <br />
  <Segment
    padded="very"
    size="mini"
    textAlign="left"
  >
    <Header
      as="h2"
      textAlign="center"
    >
      Todo list 
    </Header>
    <TodoInput />
    <TodoList
      activeItem="all"
      sortByDateAndCompleted={[Function]}
      todos={Array []}
    />
    <Footer
      handleActiveItemChange={[Function]}
      todos={Array []}
    />
  </Segment>
</div>
`;
